File: templates/problems/14_dp_2d/08_distinct_subsequences/solution.go
package distinct_subsequences

// solveNumDistinct returns the number of distinct subsequences of s that equal t. O(m*n) time, O(n) space.
func solveNumDistinct(s string, t string) int {
	m, n := len(s), len(t)
	if m < n {
		return 0
	}
	dp := make([]int, n+1)
	dp[0] = 1
	for i := 1; i <= m; i++ {
		// iterate backwards to avoid overwriting values we still need
		for j := min(i, n); j >= 1; j-- {
			if s[i-1] == t[j-1] {
				dp[j] += dp[j-1]
			}
		}
	}
	return dp[n]
}

File: templates/problems/14_dp_2d/08_distinct_subsequences/python/solution.py
def solve_num_distinct(s: str, t: str) -> int:
    """Return the number of distinct subsequences of s that equal t.

    Time: O(m*n), Space: O(m*n)
    """
    m, n = len(s), len(t)
    if m < n:
        return 0
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(m + 1):
        dp[i][0] = 1
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s[i - 1] == t[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]
            else:
                dp[i][j] = dp[i - 1][j]
    return dp[m][n]
